src/Utilities/FileSystem.cpp

```diff
@@ -0,0 +1,291 @@
+    // We used to wait no matter what before even checking for success.
+
+    // the_errno == EEXIST means that the directory exists already
+    if (status == 0 or the_errno == EEXIST) {
+      break;
+    }
+
+    // if a part of the path is missing, try walking up the path and create it
+    if (status == -1 and the_errno == ENOENT) {
+      const size_t slash = dir.rfind('/');
+      // slash is found
+      if (slash != std::string::npos) {
+        // neither trailing nor root
+        if (slash != dir.size() - 1 and slash != 0) {
+          const std::string dir2 = dir.substr(0, slash);
+          // Recursive call
+          if (tried_to_create_parent_dir) {
+            // we already tried to create this parent directory, so we must
+            // be stuck in an infinite loop, there's no point in trying further
+            // LCOV_EXCL_START
+            got_stuck_in_infinite_loop = true;
+            break;
+            // LCOV_EXCL_STOP
+          }
+          create_directory(dir2, wait_time, num_tries);
+          tried_to_create_parent_dir = true;
+          continue;
+        }
+      }
+    }
+
+    // Ok, we failed. Try again after waiting
+    // LCOV_EXCL_START
+    ++number_of_failures;
+    Parallel::printf(
+        "create_directory: mkdir(%s) failed %d time(s). Error: %d\n", dir,
+        number_of_failures, the_errno);
+    std::this_thread::sleep_for(std::chrono::duration<double>(
+        wait_time * std::pow(wait_time_increase_factor, number_of_failures)));
+    // LCOV_EXCL_STOP
+  }
+
+  // Termination conditions
+  if (terminate_on_failure) {
+    if (number_of_failures >= num_tries) {
+      // LCOV_EXCL_START
+      ERROR("Unable to mkdir '" << dir << "'. Giving up after "
+                                << number_of_failures << " tries\n");
+      // LCOV_EXCL_STOP
+    }
+    if (got_stuck_in_infinite_loop) {
+      // LCOV_EXCL_START
+      ERROR("Got stuck in an infinite loop while trying to create '" << dir
+                                                                     << "'.\n");
+      // LCOV_EXCL_STOP
+    }
+  }
+  return number_of_failures < num_tries and not got_stuck_in_infinite_loop;
+}
+
+bool check_if_dir_exists(const std::string& dir) {
+  struct stat buf {};
+  // stat returns 0 if the operation is successful (thing exists)
+  return 0 == stat(dir.c_str(), &buf) and S_ISDIR(buf.st_mode);
+}
+
+bool check_if_file_exists(const std::string& file) {
+  struct stat buf {};
+  // stat returns 0 if the operation is successful (thing exists)
+  return 0 == stat(file.c_str(), &buf) and not S_ISDIR(buf.st_mode);
+}
+
+bool is_file(const std::string& file) {
+  struct stat buf {};
+  // stat returns 0 if the operation is successful (thing exists)
+  if (0 != stat(file.c_str(), &buf)) {
+    ERROR(
+        "Failed to check if path points to a file because the path is invalid. "
+        "Given path is: "
+        << file);
+  }
+  return S_ISREG(buf.st_mode);
```

Comment:
`S_ISREG` seems reasonable, but saying that it excludes symlinks would be misleading, since `stat` resolves symlinks and will report a symlink to a regular file as a regular file (compare `lstat`).  Of course, a _broken_ symlink will be reported as a symlink.

Something like "is a regular file (or a link to one) (see stat(2))" would be my recommendation, preferably rewritten to look less like lisp.